product/src/main/java/com/yas/product/service/ProductService.java

```diff
@@ -111,73 +111,98 @@ public ProductService(ProductRepository productRepository,
+        validateProductVm(productPostVm);
 
-    private static void setValuesForVariantExisting(
-            List<ProductImage> newProductImages,
-            ProductVariationPutVm variant,
-            Product variantInDb
-    ) {
-        if (variantInDb != null) {
-            variantInDb.setName(variant.name());
-            variantInDb.setThumbnailMediaId(variant.thumbnailMediaId());
-            variantInDb.setSlug(variant.slug().toLowerCase());
-            variantInDb.setSku(variant.sku());
-            variantInDb.setGtin(variant.gtin());
-            variantInDb.setPrice(variant.price());
-            List<ProductImage> productImages = variantInDb.getProductImages();
-            if (CollectionUtils.isNotEmpty(variant.productImageIds())) {
-                variant.productImageIds().forEach(imageId -> {
-                    if (productImages.stream().noneMatch(productImage -> imageId.equals(productImage.getImageId()))) {
-                        newProductImages.add(ProductImage.builder()
-                                .imageId(imageId).product(variantInDb).build());
-                    }
-                });
-            }
+        Product mainProduct = Product.builder()
+            .name(productPostVm.name())
+            .thumbnailMediaId(productPostVm.thumbnailMediaId())
+            .slug(productPostVm.slug())
+            .description(productPostVm.shortDescription())
+            .shortDescription(productPostVm.description())
+            .specification(productPostVm.specification())
+            .sku(productPostVm.sku())
+            .gtin(productPostVm.gtin())
+            .price(productPostVm.price())
+            .isAllowedToOrder(productPostVm.isAllowedToOrder())
+            .isPublished(productPostVm.isPublished())
+            .isFeatured(productPostVm.isFeatured())
+            .isVisibleIndividually(productPostVm.isVisibleIndividually())
+            .stockTrackingEnabled(productPostVm.stockTrackingEnabled())
+            .metaTitle(productPostVm.metaTitle())
+            .metaKeyword(productPostVm.metaKeyword())
+            .metaDescription(productPostVm.description())
+            .hasOptions(CollectionUtils.isNotEmpty(productPostVm.variations())
+                && CollectionUtils.isNotEmpty(productPostVm.productOptionValues()))
+            .productCategories(List.of())
+            .taxClassId(productPostVm.taxClassId())
+            .build();
+
+        setProductBrand(productPostVm.brandId(), mainProduct);
+
+        List<ProductCategory> productCategories = setProductCategories(productPostVm.categoryIds(), mainProduct);
+
+        List<ProductImage> productImages = setProductImages(productPostVm.productImageIds(), mainProduct);
+
+        Product savedMainProduct = productRepository.saveAndFlush(mainProduct);
+        productImageRepository.saveAllAndFlush(productImages);
+        productCategoryRepository.saveAllAndFlush(productCategories);
```

Comment:
We should call the save method right after the entity is fully built (no further modification) to keep the flow straightforward. Please review and adjust if necessary